src/main/java/com/googleintern/wfm/ruleengine/action/GroupByWorkgroupId.java

```diff
@@ -0,0 +1,96 @@
+package src.main.java.com.googleintern.wfm.ruleengine.action;
+
+import com.google.common.collect.*;
+import src.main.java.com.googleintern.wfm.ruleengine.model.FilterModel;
+import src.main.java.com.googleintern.wfm.ruleengine.model.PoolAssignmentModel;
+import src.main.java.com.googleintern.wfm.ruleengine.model.RuleModel;
+import src.main.java.com.googleintern.wfm.ruleengine.model.UserPoolAssignmentModel;
+import java.util.Set;
+
+/**
+ * GroupByWorkgroupId class is used to group data by their workgroup Id value and generate possible
+ * general rules for the same workgroup.
+ */
+public class GroupByWorkgroupId {
+  /**
+   * Group data by workgroup Id.
+   * @param validData
+   * @return
+   */
+  public static ImmutableListMultimap<Long, UserPoolAssignmentModel> groupByWorkGroupId(
+      ImmutableList<UserPoolAssignmentModel> validData) {
+    ImmutableListMultimap.Builder<Long, UserPoolAssignmentModel> mapByWorkGroupIdBuilder =
+        ImmutableListMultimap.builder();
+    for (UserPoolAssignmentModel data : validData) {
+      mapByWorkGroupIdBuilder.put(data.workgroupId(), data);
+    }
+    return mapByWorkGroupIdBuilder.build();
+  }
+
+  /**
+   * Generate rules that can apply to all users from the same workgroup Id.
+   * @param mapByWorkgroupId
+   * @param workgroupId
+   * @return
+   */
+  public static ImmutableSet<RuleModel> generalRuleByWorkgroupId(
+      ImmutableListMultimap<Long, UserPoolAssignmentModel> mapByWorkgroupId, Long workgroupId) {
+    ImmutableList<UserPoolAssignmentModel> userFromSameWorkGroupId =
+        mapByWorkgroupId.get(workgroupId);
+    ImmutableSet<PoolAssignmentModel> permissionIntersections =
+        findPermissionIntersection(userFromSameWorkGroupId);
+    if (permissionIntersections == null) {
+      return ImmutableSet.of();
+    }
+    ImmutableSetMultimap<Long, Long> permissionGroup =
+        groupPermissionByCasePoolId(permissionIntersections);
+    Long workforceId = userFromSameWorkGroupId.get(0).workforceId();
+    return createGeneralRuleForWorkgroupId(permissionGroup, workforceId, workgroupId);
+  }
```

Comment:
Is this engine logic?
If so, please roll this back and save it for another commit to reduce scope for this change.